usr/src/man/man1/tabs.1
.\"
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for
.\" permission to reproduce portions of its copyrighted documentation.
.\" Original documentation from The Open Group can be obtained online at
.\" http://www.opengroup.org/bookstore/.
.\"
.\" The Institute of Electrical and Electronics Engineers and The Open
.\" Group, have given us permission to reprint portions of their
.\" documentation.
.\"
.\" In the following statement, the phrase ``this text'' refers to portions
.\" of the system documentation.
.\"
.\" Portions of this text are reprinted and reproduced in electronic form
.\" in the SunOS Reference Manual, from IEEE Std 1003.1, 2004 Edition,
.\" Standard for Information Technology -- Portable Operating System
.\" Interface (POSIX), The Open Group Base Specifications Issue 6,
.\" Copyright (C) 2001-2004 by the Institute of Electrical and Electronics
.\" Engineers, Inc and The Open Group.  In the event of any discrepancy
.\" between these versions and the original IEEE and The Open Group
.\" Standard, the original IEEE and The Open Group Standard is the referee
.\" document.  The original Standard can be obtained online at
.\" http://www.opengroup.org/unix/online.html.
.\"
.\" This notice shall appear on any product containing this material.
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\"
.\" Copyright 1989 AT&T
.\" Copyright (c) 1992, X/Open Company Limited.  All Rights Reserved.
.\" Portions Copyright (c) 1995, Sun Microsystems, Inc.  All Rights Reserved
.\"
.TH TABS 1 "Feb 1, 1995"
.SH NAME
tabs \- set tabs on a terminal
.SH SYNOPSIS
.LP
.nf
\fBtabs\fR [\fB-n\fR | \(mi\(mi\fIfile\fR
      [[\fB-code\fR] | \fB-a\fR | \fB-a2\fR | \fB-c\fR | \fB-c2\fR | \fB-c3\fR | \fB-f\fR | \fB-p\fR | \fB-s\fR | \fB-u\fR]]
      q!! [+m [\fIn\fR]] [\fB-T\fR \fItype\fR]
.fi

.LP
.nf
\fBtabs\fR [\fB-T\fR \fItype\fR] [+ \fIm\fR [\fIn\fR]] \fIn1\fR [, \fIn2\fR ,...]
.fi

.SH DESCRIPTION
.sp
.LP
The \fBtabs\fR utility sets the tab stops on the user's terminal according to a
tab specification, after clearing any previous settings. The user's terminal
must have remotely settable hardware tabs.
.SH OPTIONS
.sp
.LP
The following options are supported. If a given flag occurs more than once, the
last value given takes effect:
.sp
.ne 2
.na
\fB\fB-T\fR \fItype\fR \fR
.ad
.RS 12n
\fBtabs\fR needs to know the type of terminal in order to set tabs and margins.
\fItype\fR is a name listed in \fBterm\fR(7). If no \fB-T\fR flag is supplied,
\fBtabs\fR uses the value of the environment variable \fBTERM\fR. If the value
of \fBTERM\fR is \fINULL\fR or \fBTERM\fR is not defined in the environment
(see \fBenviron\fR(7)), \fBtabs\fR uses \fBansi+tabs\fR as the terminal type to
provide a sequence that will work for many terminals.
.RE

.sp
.ne 2
.na
\fB\fB+m[\fR\fIn\fR\fB]\fR \fR
.ad
.RS 12n
The margin argument may be used for some terminals. It causes all tabs to be
moved over \fIn\fR columns by making column \fIn\fR\fB+1\fR the left margin.
If \fB+m\fR is given without a value of \fIn\fR, the value assumed is \fB10\fR.
For a TermiNet, the first value in the tab list should be \fB1\fR, or the
margin will move even further to the right. The normal (leftmost) margin on
most terminals is obtained by \fB+m0\fR. The margin for most terminals is reset
only when the \fB+m\fR flag is given explicitly.
.RE

.SS "Tab Specification"
.sp
.LP
Four types of tab specification are accepted.  They are described below:
canned, repetitive (\fB-\fR\fIn\fR), arbitrary (\fIn1,n2,...\fR), and file
(\fB-\fR\fIfile\fR).
.sp
.LP
If no tab specification is given, the default value is \fB\(mi8\fR, that is,
UNIX system ``standard'' tabs. The lowest column number is \fB1\fR. Note:  For
\fBtabs\fR, column 1 always refers to the leftmost column on a terminal, even
one whose column markers begin at 0, for example, the DASI 300, DASI 300s, and
DASI 450.
.SS "\fICanned\fR \fB-code\fR"
.sp
.LP
Use one of the codes listed below to select a canned set of tabs. If more than
one code is specified, the last code option will be used.  The legal codes and
their meanings are as follows:
.sp
.ne 2
.na
\fB\fB-a\fR\fR
.ad
.RS 7n
\fB1,10,16,36,72\fR Assembler, IBM S/370, first format
.RE

.sp
.ne 2
.na
\fB\fB-a2\fR\fR
.ad
.RS 7n
\fB1,10,16,40,72\fR
.sp
Assembler, IBM S/370, second format
.RE

.sp
.ne 2
.na
\fB\fB-c\fR\fR
.ad
.RS 7n
\fB1,8,12,16,20,55\fR
.sp
COBOL, normal format
.RE

.sp
.ne 2
.na
\fB\fB-c2\fR\fR
.ad
.RS 7n
\fB1,6,10,14,49\fR
.sp
COBOL compact format (columns 1-6 omitted). Using this code, the first typed
character corresponds to card column 7, one space gets you to column 8, and a
tab reaches column 12. Files using this tab setup should include a format
specification as follows (see \fBfspec\fR(5)):
.sp
.in +2
.nf
<:t\fB-c2\fR \|m6 \|s66 \|d:>
.fi
.in -2

.RE

.sp
.ne 2
.na
\fB\fB-c3\fR\fR
.ad
.RS 7n
\fB1,6,10,14,18,22,26,30,34,38,42,46,50,54,58,62,67\fR
.sp
COBOL compact format (columns 1-6 omitted), with more tabs than
\fB\fR\fB-c2\fR\fB\&.\fR This is the recommended format for COBOL. The
appropriate format specification is (see \fBfspec\fR(5)):
.sp
.in +2
.nf
<:t\fB-c3\fR \|m6 \|s66 \|d:>
.fi
.in -2

.RE

.sp
.ne 2
.na
\fB\fB-f\fR\fR
.ad
.RS 7n
\fB1,7,11,15,19,23\fR
.sp
FORTRAN
.RE

.sp
.ne 2
.na
\fB\fB-p\fR\fR
.ad
.RS 7n
\fB1,5,9,13,17,21,25,29,33,37,41,45,49,53,57,61\fR
.sp
PL/I
.RE

.sp
.ne 2
.na
\fB\fB-s\fR\fR
.ad
.RS 7n
\fB1,10,55\fR
.sp
SNOBOL
.RE

.sp
.ne 2
.na
\fB\fB-u\fR\fR
.ad
.RS 7n
\fB1,12,20,44\fR
.sp
UNIVAC 1100 Assembler
.RE

.SS "\fIRepetitive\fR"
.sp
.ne 2
.na
\fB\fB-\fR\fIn\fR \fR
.ad
.RS 7n
A \fIrepetitive\fR specification requests tabs at columns 1+\fIn\fR,
1+2*\fIn\fR, etc., where \fIn\fR is a single-digit decimal number. Of
particular importance is the value \fB8\fR: this represents the UNIX system
``standard'' tab setting, and is the most likely tab setting to be found at a
terminal. When \fB\(mi0\fR is used, the tab stops are cleared and no new ones
are set.
.RE

.SS "\fIArbitrary\fR"
.sp
.LP
See OPERANDS.
.SS "\fIFile\fR"
.sp
.ne 2
.na
\fB\fB-\fR\fIfile\fR\fR
.ad
.RS 9n
If the name of a \fIfile\fR is given, \fBtabs\fR reads the first line of the
file, searching for a format specification (see \fBfspec\fR(5)). If it finds
one there, it sets the tab stops according to it, otherwise it sets them as
\fB\(mi8\fR\&. This type of specification may be used to make sure that a
tabbed file is printed with correct tab settings, and would be used with the
\fBpr\fR command:
.sp
.in +2
.nf
example% \fBtabs -\fI file\fR; pr\fI file\fR\fR
.fi
.in -2
.sp

.RE

.sp
.LP
Tab and margin setting is performed via the standard output.
.SH OPERANDS
.sp
.LP
The following operand is supported:
.sp
.ne 2
.na
\fB\fIn1\fR[,\fIn2\fR,\|.\|.\|.] \fR
.ad
.RS 22n
The \fIarbitrary\fR format consists of tab-stop values separated by commas or
spaces. The tab-stop values must be positive decimal integers in ascending
order. Up to 40 numbers are allowed. If any number (except the first one) is
preceded by a plus sign, it is taken as an increment to be added to the
previous value. Thus, the formats \fB1\fR,\fB10\fR,\fB20\fR,\fB30\fR, and
\fB1\fR,\fB10\fR,\fB+10\fR,\fB+10\fR are considered identical.
.RE

.SH EXAMPLES
.LP
\fBExample 1 \fRUsing the tabs command
.sp
.LP
The following command is an example using \fB-code\fR ( \fIcanned\fR
specification) to set tabs to the settings required by the IBM assembler:
columns 1, 10, 16, 36, 72:

.sp
.in +2
.nf
example% \fBtabs -a\fR
.fi
.in -2
.sp

.sp
.LP
The next command is an example of using \fB-n\fR (\fIrepetitive\fR
specification), where \fIn\fR is \fB8\fR, causes tabs to be set every eighth
position: 1+(1*8), 1+(2*8), .\|.\|.\| which evaluate to columns 9,
17, .\|.\|.\|:

.sp
.in +2
.nf
example% \fBtabs \(mi8\fR
.fi
.in -2
.sp

.sp
.LP
This command uses \fIn1\fR,\fIn2\fR,.\|.\|.\| (\fIarbitrary\fR specification)
to set tabs at columns 1, 8, and 36:

.sp
.in +2
.nf
example% \fBtabs 1,8,36\fR
.fi
.in -2
.sp

.sp
.LP
The last command is an example of using \fI-file\fR (\fBfile\fR specification)
to indicate that tabs should be set according to the first line of
$\fBHOME\fR/fspec.list/att4425  (see \fBfspec\fR(5)).

.sp
.in +2
.nf
example% \fBtabs -$HOME/fspec.list/att4425\fR
.fi
.in -2
.sp

.SH ENVIRONMENT VARIABLES
.sp
.LP
See \fBenviron\fR(7) for descriptions of the following environment variables
that affect the execution of \fBtabs\fR: \fBLANG\fR, \fBLC_ALL\fR,
\fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.sp
.ne 2
.na
\fB\fBTERM\fR \fR
.ad
.RS 9n
Determine the terminal type. If this variable is unset or null, and if the
\fB-T\fR option is not specified, terminal type \fBansi+tabs\fR will be used.
.RE

.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.na
\fB\fB0\fR \fR
.ad
.RS 7n
Successful completion.
.RE

.sp
.ne 2
.na
\fB\fB>0\fR \fR
.ad
.RS 7n
An error occurred.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(7) for descriptions of the following attributes:
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
CSI	Enabled
_
Interface Stability	Standard
.TE

.SH SEE ALSO
.sp
.LP
.BR expand (1),
.BR newform (1),
.BR pr (1),
.BR stty (1),
.BR tput (1),
.BR fspec (5),
.BR terminfo (5),
.BR attributes (7),
.BR environ (7),
.BR standards (7),
.BR term (7)
.SH NOTES
.sp
.LP
There is no consistency among different terminals regarding ways of clearing
tabs and setting the left margin.
.sp
.LP
\fBtabs\fR clears only \fB20\fR tabs (on terminals requiring a long sequence),
but is willing to set \fB64\fR.
.sp
.LP
The \fItabspec\fR used with the \fBtabs\fR command is different from the one
used with the \fBnewform\fR command. For example, \fBtabs\fR \fB\(mi8\fR sets
every eighth position; whereas \fBnewform\fR \fB\(mii\(mi8\fR indicates that
tabs are set every eighth position.


usr/src/cmd/tabs/tabs.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 1996 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/

#pragma ident	"%Z%%M%	%I%	%E% SMI"

/*
 *	tabs [tabspec] [+mn] [-Ttype]
 *	set tabs (and margin, if +mn), for terminal type
 */


#include <stdio.h>
#include <signal.h>
#include <sys/types.h>
#include <stdlib.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <curses.h>
#include <term.h>
#include <locale.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <limits.h>
#include <signal.h>

#define	EQ(a, b)	(strcmp(a, b) == 0)
/*	max # columns used (needed for GSI) */
#define	NCOLS	256
#define	NTABS	65	/* max # tabs +1 (to be set) */
#define	NTABSCL	21	/* max # tabs + 1 that will be cleared */
#define	ESC	033
#define	CLEAR	'2'
#define	SET	'1'
#define	TAB	'\t'
#define	CR	'\r'
#define	NMG	0	/* no margin setting */
#define	GMG	1	/* DTC300s margin */
#define	TMG	2	/* TERMINET margin */
#define	DMG	3	/* DASI450 margin */
#define	FMG	4	/* TTY 43 margin */
#define	TRMG	5	/* Trendata 4000a */

#define	TCLRLN	0	/* long, repetitive, general tab clear */

static char tsethp[] = {ESC,  '1', 0};		/* (default) */
static char tsetibm[] = {ESC, '0', 0};		/* ibm */
static char tclrhp[] = {ESC, '3', CR, 0};	/* hp terminals */
/* short sequence for many terminals */
static char tclrsh[] = {ESC, CLEAR, CR, 0};
static char tclrgs[] = {ESC, TAB, CR, 0};	/* short, for 300s */
static char tclr40[] = {ESC, 'R', CR, 0};	/* TTY 40/2, 4424 */
static char tclribm[] = {ESC, '1', CR, 0};	/* ibm */

static struct ttab {
	char *ttype;	/* -Tttype */
	char *tclr;	/* char sequence to clear tabs and return carriage */
	int tmaxtab;	/* maximum allowed position */
} *tt;

static struct ttab termtab[] = {
	"",		tclrsh,	132,
	"1620-12",	tclrsh,	158,
	"1620-12-8",	tclrsh,	158,
	"1700-12",	tclrsh,	132,
	"1700-12-8",	tclrsh,	158,
	"300-12",	TCLRLN,	158,
	"300s-12",	tclrgs,	158,
	"4424",		tclr40,	 80,
	"4000a",	tclrsh,	132,
	"4000a-12",	tclrsh,	158,
	"450-12",	tclrsh,	158,
	"450-12-8",	tclrsh,	158,
	"2631",		tclrhp, 240,
	"2631-c",	tclrhp, 240,
	"ibm",		tclribm, 80,
	0
};

static int	err;
static int 	tmarg;
static char	settab[32], clear_tabs[32];

static int	maxtab;		/* max tab for repetitive spec */
static int	margin;
static int	margflg;	/* >0 ==> +m option used, 0 ==> not */
static char	*terminal = "";
static char	*tabspec = "-8";	/* default tab specification */

static struct termio ttyold;	/* tty table */
static int	ttyisave;	/* save for input modes */
static int	ttyosave;	/* save for output modes */
static int	istty;		/* 1 ==> is actual tty */

static struct	stat	statbuf;
static char	*devtty;

static void scantab(char *scan, int tabvect[NTABS], int level);
static void repetab(char *scan, int tabvect[NTABS]);
static void arbitab(char *scan, int tabvect[NTABS]);
static void filetab(char *scan, int tabvect[NTABS], int level);
static int getmarg(char *term);
static struct ttab *termadj();
static void settabs(int tabvect[NTABS]);
static char *cleartabs(register char *p, char *qq);
static int getnum(char **scan1);
static void endup();
static int stdtab(char option[], int tabvect[]);
static void usage();
static int chk_codes(char *codes);

int
main(int argc, char **argv)
{
	int tabvect[NTABS];	/* build tab list here */
	char *scan;	/* scan pointer to next char */
	char operand[LINE_MAX];
	int option_end = 0;

	(void) setlocale(LC_ALL, "");

#if !defined(TEXT_DOMAIN)
#define	TEXT_DOMAIN "SYS_TEST"
#endif
	(void) textdomain(TEXT_DOMAIN);

	(void) signal(SIGINT, endup);
	if (ioctl(1, TCGETA, &ttyold) == 0) {
		ttyisave = ttyold.c_iflag;
		ttyosave = ttyold.c_oflag;
		(void) fstat(1, &statbuf);
		devtty = ttyname(1);
		(void) chmod(devtty, 0000);	/* nobody, not even us */
		istty++;
	}
	tabvect[0] = 0;	/* mark as not yet filled in */
	while (--argc > 0) {
		scan = *++argv;
		if (*scan == '+') {
			if (!option_end) {
				if (*++scan == 'm') {
					margflg++;
					if (*++scan)
						margin = getnum(&scan);
					else
						margin = 10;
				} else {
					(void) fprintf(stderr, gettext(
				"tabs: %s: invalid tab spec\n"), scan-1);
					usage();
				}
			} else {
				/*
				 * only n1[,n2,...] operand can follow
				 * end of options delimiter "--"
				 */
				(void) fprintf(stderr, gettext(
				"tabs: %s: invalid tab stop operand\n"), scan);
				usage();
			}
		} else if (*scan == '-') {
			if (!option_end) {
				if (*(scan+1) == 'T') {
					/* allow space or no space after -T */
					if (*(scan+2) == '\0') {
						if (--argc > 0)
							terminal = *++argv;
						else
							usage();
					} else
						terminal = scan+2;
				} else if (*(scan+1) == '-')
					if (*(scan+2) == '\0')
						option_end = 1;
					else
						tabspec = scan; /* --file */
				else if (strcmp(scan+1, "code") == 0) {
					/* EMPTY */
					/* skip to next argument */
				} else if (chk_codes(scan+1) ||
				    (isdigit(*(scan+1)) && *(scan+2) == '\0')) {
					/*
					 * valid code or single digit decimal
					 * number
					 */
					tabspec = scan;
				} else {
					(void) fprintf(stderr, gettext(
					"tabs: %s: invalid tab spec\n"), scan);
					usage();
				}
			} else {
				/*
				 * only n1[,n2,...] operand can follow
				 * end of options delimiter "--"
				 */
				(void) fprintf(stderr, gettext(
				"tabs: %s: invalid tab stop operand\n"), scan);
				usage();
			}
		} else {
			/*
			 * Tab-stop values separated using either commas
			 * or blanks.  If any number (except the first one)
			 * is preceded by a plus sign, it is taken as an
			 * increment to be added to the previous value.
			 */
			operand[0] = '\0';
			while (argc > 0) {
				if (strrchr(*argv, '-') == (char *)NULL) {
					(void) strcat(operand, *argv);
					if (argc > 1)
						(void) strcat(operand, ",");
					--argc;
					++argv;
				} else {
					(void) fprintf(stderr, gettext(
		"tabs: %s: tab stop values must be positive integers\n"),
					    *argv);
					usage();
				}
			}
			tabspec = operand;	/* save tab specification */
		}
	}
	if (*terminal == '\0') {
		if ((terminal = getenv("TERM")) == (char *)NULL ||
		    *terminal == '\0') {
			/*
			 * Use tab setting and clearing sequences specified
			 * by the ANSI standard.
			 */
			terminal = "ansi+tabs";
		}
	}
	if (setupterm(terminal, 1, &err) == ERR) {
		(void) fprintf(stderr, gettext(
		"tabs: %s: terminfo file not found\n"), terminal);
		usage();
	} else if (!tigetstr("hts")) {
		(void) fprintf(stderr, gettext(
		"tabs: cannot set tabs on terminal type %s\n"), terminal);
		usage();
	}
	if (err <= 0 || columns <= 0 || set_tab == 0) {
		tt = termadj();
		if (strcmp(terminal, "ibm") == 0)
			(void) strcpy(settab, tsetibm);
		else
			(void) strcpy(settab, tsethp);
		(void) strcpy(clear_tabs, tt->tclr);
		maxtab = tt->tmaxtab;
	} else {
		maxtab = columns;
		(void) strcpy(settab, set_tab);
		(void) strcpy(clear_tabs, clear_all_tabs);
	}
	scantab(tabspec, tabvect, 0);
	if (!tabvect[0])
		repetab("8", tabvect);
	settabs(tabvect);
	endup();
	return (0);
}

/*
 * return 1 if code option is valid, otherwise return 0
 */
int
chk_codes(char *code)
{
	if (*(code+1) == '\0' && (*code == 'a' || *code == 'c' ||
	    *code == 'f' || *code == 'p' || *code == 's' || *code == 'u'))
			return (1);
	else if (*(code+1) == '2' && *(code+2) == '\0' &&
	    (*code == 'a' || *code == 'c'))
			return (1);
	else if (*code == 'c' && *(code+1) == '3' && *(code+2) == '\0')
		return (1);
	return (0);
}

/*	scantab: scan 1 tabspec & return tab list for it */
void
scantab(char *scan, int tabvect[NTABS], int level)
{
	char c;
	if (*scan == '-') {
		if ((c = *++scan) == '-')
			filetab(++scan, tabvect, level);
		else if (c >= '0' && c <= '9')
			repetab(scan, tabvect);
		else if (stdtab(scan, tabvect)) {
			endup();
			(void) fprintf(stderr, gettext(
			"tabs: %s: unknown tab code\n"), scan);
			usage();
		}
	} else {
		arbitab(scan, tabvect);
	}
}

/*	repetab: scan and set repetitve tabs, 1+n, 1+2*n, etc */

void
repetab(char *scan, int tabvect[NTABS])
{
	int incr, i, tabn;
	int limit;
	incr = getnum(&scan);
	tabn = 1;
	limit = (maxtab-1)/(incr?incr:1)-1; /* # last actual tab */
	if (limit > NTABS-2)
		limit = NTABS-2;
	for (i = 0; i <= limit; i++)
		tabvect[i] = tabn += incr;
	tabvect[i] = 0;
}

/*	arbitab: handle list of arbitrary tabs */

void
arbitab(char *scan, int tabvect[NTABS])
{
	char *scan_save;
	int i, t, last;

	scan_save = scan;
	last = 0;
	for (i = 0; i < NTABS-1; ) {
		if (*scan == '+') {
			scan++;		/* +n ==> increment, not absolute */
			if (t = getnum(&scan))
				tabvect[i++] = last += t;
			else {
				endup();
				(void) fprintf(stderr, gettext(
				"tabs: %s: invalid increment\n"), scan_save);
				usage();
			}
		} else {
			if ((t = getnum(&scan)) > last)
				tabvect[i++] = last = t;
			else {
				endup();
				(void) fprintf(stderr, gettext(
				"tabs: %s: invalid tab stop\n"), scan_save);
				usage();
			}
		}
		if (*scan++ != ',') break;
	}
	if (last > NCOLS) {
		endup();
		(void) fprintf(stderr, gettext(
	"tabs: %s: last tab stop would be set at a column greater than %d\n"),
		    scan_save, NCOLS);
		usage();
	}
	tabvect[i] = 0;
}

/*	filetab: copy tabspec from existing file */
#define	CARDSIZ	132

void
filetab(char *scan, int tabvect[NTABS], int level)
{
	int length, i;
	char c;
	int fildes;
	char card[CARDSIZ];	/* buffer area for 1st card in file */
	char state, found;
	char *temp;
	if (level) {
		endup();
		(void) fprintf(stderr, gettext(
		"tabs: %s points to another file: invalid file indirection\n"),
		    scan);
		exit(1);
	}
	if ((fildes = open(scan, O_RDONLY)) < 0) {
		endup();
		(void) fprintf(stderr, gettext("tabs: %s: "), scan);
		perror("");
		exit(1);
	}
	length = read(fildes, card, CARDSIZ);
	(void) close(fildes);
	found = state = 0;
	scan = 0;
	for (i = 0; i < length && (c = card[i]) != '\n'; i++) {
		switch (state) {
		case 0:
			state = (c == '<'); break;
		case 1:
			state = (c == ':')?2:0; break;
		case 2:
			if (c == 't')
				state = 3;
			else if (c == ':')
				state = 6;
			else if (c != ' ')
				state = 5;
			break;
		case 3:
			if (c == ' ')
				state = 2;
			else {
				scan = &card[i];
				state = 4;
			}
			break;
		case 4:
			if (c == ' ') {
				card[i] = '\0';
				state = 5;
			} else if (c == ':') {
				card[i] = '\0';
				state = 6;
			}
			break;
		case 5:
			if (c == ' ')
				state = 2;
			else if (c == ':')
				state = 6;
			break;
		case 6:
			if (c == '>') {
				found = 1;
				goto done;
			} else state = 5;
			break;
		}
	}
done:
	if (found && scan != 0) {
		scantab(scan, tabvect, 1);
		temp = scan;
		while (*++temp)
			;
		*temp = '\n';
	}
	else
		scantab("-8", tabvect, 1);
}

int
getmarg(char *term)
{
	if (strncmp(term, "1620", 4) == 0 ||
	    strncmp(term, "1700", 4) == 0 || strncmp(term, "450", 3) == 0)
		return (DMG);
	else if (strncmp(term, "300s", 4) == 0)
		return (GMG);
	else if (strncmp(term, "4000a", 5) == 0)
		return (TRMG);
	else if (strcmp(term, "43") == 0)
		return (FMG);
	else if (strcmp(term, "tn300") == 0 || strcmp(term, "tn1200") == 0)
		return (TMG);
	else
		return (NMG);
}



struct ttab *
termadj(void)
{
	struct ttab *t;

	if (strncmp(terminal, "40-2", 4) == 0 || strncmp(terminal,
	    "40/2", 4) == 0 || strncmp(terminal, "4420", 4) == 0)
		(void) strcpy(terminal, "4424");
	else if (strncmp(terminal, "ibm", 3) == 0 || strcmp(terminal,
	    "3101") == 0 || strcmp(terminal, "system1") == 0)
		(void) strcpy(terminal, "ibm");

	for (t = termtab; t->ttype; t++) {
		if (EQ(terminal, t->ttype))
			return (t);
	}
/* should have message */
	return (termtab);
}

char	*cleartabs();
/*
 *	settabs: set actual tabs at terminal
 *	note: this code caters to necessities of handling GSI and
 *	other terminals in a consistent way.
 */

void
settabs(int tabvect[NTABS])
{
	char setbuf[512];	/* 2+3*NTABS+2+NCOLS+NTABS (+ some extra) */
	char *p;		/* ptr for assembly in setbuf */
	int *curtab;		/* ptr to tabvect item */
	int i, previous, nblanks;
	if (istty) {
		ttyold.c_iflag &= ~ICRNL;
		ttyold.c_oflag &= ~(ONLCR|OCRNL|ONOCR|ONLRET);
		(void) ioctl(1, TCSETAW, &ttyold);	/* turn off cr-lf map */
	}
	p = setbuf;
	*p++ = CR;
	p = cleartabs(p, clear_tabs);

	if (margflg) {
		tmarg = getmarg(terminal);
		switch (tmarg) {
		case GMG:	/* GSI300S */
		/*
		 * NOTE: the 300S appears somewhat odd, in that there is
		 * a column 0, but there is no way to do a direct tab to it.
		 * The sequence ESC 'T' '\0' jumps to column 27 and prints
		 * a '0', without changing the margin.
		 */
			*p++ = ESC;
			*p++ = 'T';	/* setup for direct tab */
			if (margin &= 0177)	/* normal case */
				*p++ = margin;
			else {			/* +m0 case */
				*p++ = 1;	/* column 1 */
				*p++ = '\b';	/* column 0 */
			}
			*p++ = margin;	/* direct horizontal tab */
			*p++ = ESC;
			*p++ = '0';	/* actual margin set */
			break;
		case TMG:	/* TERMINET 300 & 1200 */
			while (margin--)
				*p++ = ' ';
			break;
		case DMG:	/* DASI450/DIABLO 1620 */
			*p++ = ESC;	/* direct tab ignores margin */
			*p++ = '\t';
			if (margin == 3) {
				*p++ = (margin & 0177);
				*p++ = ' ';
			}
			else
				*p++ = (margin & 0177) + 1;
			*p++ = ESC;
			*p++ = '9';
			break;
		case FMG:	/* TTY 43 */
			p--;
			*p++ = ESC;
			*p++ = 'x';
			*p++ = CR;
			while (margin--)
				*p++ = ' ';
			*p++ = ESC;
			*p++ = 'l';
			*p++ = CR;
			(void) write(1, setbuf, p - setbuf);
			return;
		case TRMG:
			p--;
			*p++ = ESC;
			*p++ = 'N';
			while (margin--)
				*p++ = ' ';
			*p++ = ESC;
			*p++ = 'F';
			break;
		}
	}

/*
 *	actual setting: at least terminals do this consistently!
 */
	previous = 1; curtab = tabvect;
	while ((nblanks = *curtab-previous) >= 0 &&
	    previous + nblanks <= maxtab) {
		for (i = 1; i <= nblanks; i++) *p++ = ' ';
		previous = *curtab++;
		(void) strcpy(p, settab);
		p += strlen(settab);
	}
	*p++ = CR;
	if (EQ(terminal, "4424"))
		*p++ = '\n';	/* TTY40/2 needs LF, not just CR */
	(void) write(1, setbuf, p - setbuf);
}


/*
 *	Set software tabs.  This only works on UNIX/370 using a series/1
 *	front-end processor.
 */


/*	cleartabs(pointer to buffer, pointer to clear sequence) */
char *
cleartabs(register char *p, char *qq)
{
	int i;
	char *q;
	q = qq;
	if (clear_tabs == 0) {		/* if repetitive sequence */
		*p++ = CR;
		for (i = 0; i < NTABSCL - 1; i++) {
			*p++ = TAB;
			*p++ = ESC;
			*p++ = CLEAR;
		}
		*p++ = CR;
	} else {
		while (*p++ = *q++)	/* copy table sequence */
			;
		p--;			/* adjust for null */
		if (EQ(terminal, "4424")) {	/* TTY40 extra delays needed */
			*p++ = '\0';
			*p++ = '\0';
			*p++ = '\0';
			*p++ = '\0';
		}
	}
	return (p);
}
/*	getnum: scan and convert number, return zero if none found */
/*	set scan ptr to addr of ending delimeter */
int
getnum(char **scan1)
{
	int n;
	char c, *scan;
	n = 0;
	scan = *scan1;
	while ((c = *scan++) >= '0' && c <= '9') n = n * 10 + c -'0';
	*scan1 = --scan;
	return (n);
}

/*	usage: terminate processing with usage message */
void
usage(void)
{
	(void) fprintf(stderr, gettext(
"usage: tabs [ -n| --file| [[-code] -a| -a2| -c| -c2| -c3| -f| -p| -s| -u]] \
[+m[n]] [-T type]\n"));

	(void) fprintf(stderr, gettext(
"       tabs [-T type][+m[n]] n1[,n2,...]\n"));

	endup();
	exit(1);
}

/*	endup: make sure tty mode reset & exit */
void
endup(void)
{

	if (istty) {
		ttyold.c_iflag = ttyisave;
		ttyold.c_oflag = ttyosave;
		/* reset cr-lf to previous */
		(void) ioctl(1, TCSETAW, &ttyold);
		(void) chmod(devtty, statbuf.st_mode);
	}
	if (err > 0) {
		(void) resetterm();
	}
}

/*
 *	stdtabs: standard tabs table
 *	format: option code letter(s), null, tabs, null
 */
static char stdtabs[] = {
'a',	0, 1, 10, 16, 36, 72, 0,		/* IBM 370 Assembler */
'a', '2', 0, 1, 10, 16, 40, 72, 0,		/* IBM Assembler alternative */
'c',	0, 1, 8, 12, 16, 20, 55, 0,		/* COBOL, normal */
'c', '2', 0, 1, 6, 10, 14, 49, 0,		/* COBOL, crunched */
'c', '3', 0, 1, 6, 10, 14, 18, 22, 26, 30, 34, 38, 42, 46, 50, 54, 58, 62, 67,
	0,					/* crunched COBOL, many tabs */
'f',	0, 1, 7, 11, 15, 19, 23, 0,		/* FORTRAN */
'p',	0, 1, 5, 9, 13, 17, 21, 25, 29, 33, 37, 41, 45, 49, 53, 57, 61, 0,
						/* PL/I */
's',	0, 1, 10, 55, 0, 			/* SNOBOL */
'u',	0, 1, 12, 20, 44, 0,			/* UNIVAC ASM */
0};

/*
 *	stdtab: return tab list for any "canned" tab option.
 *	entry: option points to null-terminated option string
 *		tabvect points to vector to be filled in
 *	exit: return (0) if legal, tabvect filled, ending with zero
 *		return (-1) if unknown option
 */
int
stdtab(char option[], int tabvect[])
{
	char *sp;
	tabvect[0] = 0;
	sp = stdtabs;
	while (*sp) {
		if (EQ(option, sp)) {
			while (*sp++)		/* skip to 1st tab value */
				;
			while (*tabvect++ = *sp++)	/* copy, make int */
				;
			return (0);
		}
		while (*sp++)	/* skip to 1st tab value */
			;
		while (*sp++)		/* skip over tab list */
			;
	}
	return (-1);
}
